`impl dyn Trait` on an auto trait produces a confusing diagnostic

Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=850ae3ba756280ffd028d7342926adea)):

```rust
impl dyn Unpin {}
```

The current output is:

```
error[E0118]: no nominal type found for inherent implementation
 --> src/lib.rs:1:6
  |
1 | impl dyn Unpin {}
  |      ^^^^^^^^^ impl requires a nominal type
  |
  = note: either implement a trait on it or create a newtype to wrap it instead
```

This is rather confusing. It's not immediately clear what the difference between `impl dyn Unpin` and this working code is ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c160754c43fd35dded133da1be1352ad)):
```rust
trait A {}
impl dyn A {}
```

Since `impl dyn A` is accepted, the user might be confused why `dyn A` would be considered a nominal type when `dyn Unpin` isn't. The real reason why `impl dyn Unpin` is an error is because it is an `impl` block for ~~a type~~ [edit: a dynamically dispatched auto trait] from another crate, which should be handled by [`E0116`](https://doc.rust-lang.org/error-index.html#E0116).

Ideally the output should look like:

```
error[E0116]: cannot define inherent `impl` for a type outside of the crate where the type is defined
 --> src/lib.rs:1:1
  |
1 | impl dyn Unpin {}
  | ^^^^^^^^^^^^^^^^^ impl for type defined outside of crate.
  |
  = note: define and implement a trait or new type instead
```

It seems like this only happens with auto traits. Regular traits already emit E0116 in this case ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b77ac5b86d42a5389c284cfd92dcecdd)):
```rust
use std::panic::UnwindSafe;
impl dyn Copy {}
impl dyn UnwindSafe {}
```
```
error[E0116]: cannot define inherent `impl` for a type outside of the crate where the type is defined
 --> src/lib.rs:3:1
  |
3 | impl dyn Copy {}
  | ^^^^^^^^^^^^^^^^ impl for type defined outside of crate.
  |
  = note: define and implement a trait or new type instead

error[E0118]: no nominal type found for inherent implementation
 --> src/lib.rs:4:6
  |
4 | impl dyn UnwindSafe {}
  |      ^^^^^^^^^^^^^^ impl requires a nominal type
  |
  = note: either implement a trait on it or create a newtype to wrap it instead
```
@rustbot modify labels: +D-confusing +F-auto_traits

--- a/compiler/rustc_error_codes/src/error_codes.rs
+++ b/compiler/rustc_error_codes/src/error_codes.rs
@@ -480,6 +480,7 @@ E0781: include_str!("./error_codes/E0781.md"),
 E0782: include_str!("./error_codes/E0782.md"),
 E0783: include_str!("./error_codes/E0783.md"),
 E0784: include_str!("./error_codes/E0784.md"),
+E0785: include_str!("./error_codes/E0785.md"),
 ;
 //  E0006, // merged with E0005
 //  E0008, // cannot bind by-move into a pattern guard

--- a/compiler/rustc_error_codes/src/error_codes/E0785.md
+++ b/compiler/rustc_error_codes/src/error_codes/E0785.md
@@ -26,5 +26,5 @@ trait PrincipalTrait {}
 
 auto trait AutoTrait {}
 
-impl dyn (PrincipalTrait + AutoTrait + Send) {}
+impl dyn PrincipalTrait + AutoTrait + Send {}
 ```

--- a/compiler/rustc_error_codes/src/error_codes/E0785.md
+++ b/compiler/rustc_error_codes/src/error_codes/E0785.md
@@ -1,0 +1,30 @@
+An inherent `impl` was written on a dyn auto trait.
+
+Erroneous code example:
+
+```compile_fail,E0785
+#![feature(auto_traits)]
+
+auto trait AutoTrait {}
+
+impl dyn AutoTrait {}
+```
+
+Dyn objects allow any number of auto traits, plus at most one non-auto trait.
+The non-auto trait becomes the "principal trait".
+
+When checking if an impl on a dyn trait is coherent, the principal trait is
+normally the only one considered. Since the erroneous code has no principal
+trait, it cannot be implemented at all.
+
+Working example:
+
+```
+#![feature(auto_traits)]
+
+trait PrincipalTrait {}
+
+auto trait AutoTrait {}
+
+impl dyn (PrincipalTrait + AutoTrait + Send) {}
+```

--- a/compiler/rustc_typeck/src/coherence/inherent_impls.rs
+++ b/compiler/rustc_typeck/src/coherence/inherent_impls.rs
@@ -60,6 +60,17 @@ impl ItemLikeVisitor<'v> for InherentCollect<'tcx> {
             ty::Dynamic(ref data, ..) if data.principal_def_id().is_some() => {
                 self.check_def_id(item, data.principal_def_id().unwrap());
             }
+            ty::Dynamic(..) => {
+                struct_span_err!(
+                    self.tcx.sess,
+                    ty.span,
+                    E0785,
+                    "cannot define inherent `impl` for a dyn auto trait"
+                )
+                .span_label(ty.span, "impl requires a principal trait")
+                .note("define and implement a new trait or type instead")
+                .emit();
+            }
             ty::Bool => {
                 self.check_primitive_impl(
                     item.def_id,

--- a/compiler/rustc_typeck/src/coherence/inherent_impls.rs
+++ b/compiler/rustc_typeck/src/coherence/inherent_impls.rs
@@ -67,7 +67,7 @@ impl ItemLikeVisitor<'v> for InherentCollect<'tcx> {
                     E0785,
                     "cannot define inherent `impl` for a dyn auto trait"
                 )
-                .span_label(ty.span, "impl requires a principal trait")
+                .span_label(ty.span, "impl requires at least one non-auto trait")
                 .note("define and implement a new trait or type instead")
                 .emit();
             }

--- a/src/test/ui/coherence/issue-85026.rs
+++ b/src/test/ui/coherence/issue-85026.rs
@@ -1,0 +1,10 @@
+#![feature(auto_traits)]
+auto trait AutoTrait {}
+
+// You cannot impl your own `dyn AutoTrait`.
+impl dyn AutoTrait {} //~ERROR E0785
+
+// You cannot impl someone else's `dyn AutoTrait`
+impl dyn Unpin {} //~ERROR E0785
+
+fn main() {}

--- a/src/test/ui/coherence/issue-85026.stderr
+++ b/src/test/ui/coherence/issue-85026.stderr
@@ -1,0 +1,19 @@
+error[E0785]: cannot define inherent `impl` for a dyn auto trait
+  --> $DIR/issue-85026.rs:5:6
+   |
+LL | impl dyn AutoTrait {}
+   |      ^^^^^^^^^^^^^ impl requires a principal trait
+   |
+   = note: define and implement a new trait or type instead
+
+error[E0785]: cannot define inherent `impl` for a dyn auto trait
+  --> $DIR/issue-85026.rs:8:6
+   |
+LL | impl dyn Unpin {}
+   |      ^^^^^^^^^ impl requires a principal trait
+   |
+   = note: define and implement a new trait or type instead
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0785`.

--- a/src/test/ui/coherence/issue-85026.stderr
+++ b/src/test/ui/coherence/issue-85026.stderr
@@ -2,7 +2,7 @@ error[E0785]: cannot define inherent `impl` for a dyn auto trait
   --> $DIR/issue-85026.rs:5:6
    |
 LL | impl dyn AutoTrait {}
-   |      ^^^^^^^^^^^^^ impl requires a principal trait
+   |      ^^^^^^^^^^^^^ impl requires at least one non-auto trait
    |
    = note: define and implement a new trait or type instead
 
@@ -10,7 +10,7 @@ error[E0785]: cannot define inherent `impl` for a dyn auto trait
   --> $DIR/issue-85026.rs:8:6
    |
 LL | impl dyn Unpin {}
-   |      ^^^^^^^^^ impl requires a principal trait
+   |      ^^^^^^^^^ impl requires at least one non-auto trait
    |
    = note: define and implement a new trait or type instead
 

